dpctl/tensor/libtensor/include/utils/math_utils.hpp

```diff
@@ -0,0 +1,104 @@
+//===------- math_utils.hpp - Implementation of math utils  -------*-C++-*/===//
+//
+//                      Data Parallel Control (dpctl)
+//
+// Copyright 2020-2023 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file defines math utility functions.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+#include <cmath>
+#include <complex>
+
+namespace dpctl
+{
+namespace tensor
+{
+namespace func_utils
+{
+
+template <T>
+bool less_complex(const std::complex<T> &in1, const std::complex<T> &in2)
+{
+    T real1 = std::real(in1);
+    T real2 = std::real(in2);
+    T imag1 = std::imag(in1);
+
+    return (real1 == real2) ? (imag1 < std::imag(x2)) : real1 < real2;
+}
+
+template <T>
+bool greater_complex(const std::complex<T> &in1, const std::complex<T> &in2)
+{
+    T real1 = std::real(in1);
+    T real2 = std::real(in2);
+    T imag1 = std::imag(in1);
+
+    return (real1 == real2) ? (imag1 > std::imag(x2)) : real1 > real2;
+}
+
+template <T>
+bool less_equal_complex(const std::complex<T> &in1, const std::complex<T> &in2)
+{
+    T real1 = std::real(in1);
+    T real2 = std::real(in2);
+    T imag1 = std::imag(in1);
+
+    return (real1 == real2) ? (imag1 <= std::imag(x2)) : real1 <= real2;
+}
+
+template <T>
+bool greater_equal_complex(const std::complex<T> &in1,
+                           const std::complex<T> &in2)
+{
+    T real1 = std::real(in1);
+    T real2 = std::real(in2);
+    T imag1 = std::imag(in1);
+
+    return (real1 == real2) ? (imag1 >= std::imag(x2)) : real1 >= real2;
+}
+
+template <T>
+std::complex<T> max_complex(const std::complex<T> &in1,
+                            const std::complex<T> &in2)
+{
+    T real1 = std::real(in1);
+    T real2 = std::real(in2);
+    T imag1 = std::imag(in1);
+
+    bool is_greater =
+        (real1 == real2) ? (imag1 > std::imag(x2)) : real1 > real2;
+    return (std::isnan(real1) || std::isnan(imag1) || greater_c) ? in1 : in2;
```

Comment:
Following your suggestion in #1316 , I noticed it does not return `in2` if imaginary part of `in2` is `NaN`.
Something similar to the following works fine.

```suggestion
            T imag2 = std::imag(in2);
            
            if (std::isnan(real1) || std::isnan(imag1))
                return in1;
            else if (std::isnan(real2) || std::isnan(imag2))
                return in2;
            else if (real1 == real2) 
                return imag1 > imag2 ? in1 : in2;
            else
                return real1 > real2 ? in1 : in2;
```